### ISSUE TITLE
JSON spec

### ISSUE THREAD

**Issue by ardubs** [ckolderup/picaro#3]
Had some ideas to throw out for additional things that might be useful to include in our initial spec. Some of them may be v1.5 at best, but figure it's best to get them down. What's already there is pretty expressive, so I don't think we need all that much.

For v1.0 / ASAP:
- A starting room property. Not sure if you'd want to do this as a boolean property of a room or as a global flag that points to a certain room name, but we should probably have this in there somehow.
- Don't know where this should go, but we need a way to signal that a story is over. David suggested being able to spin this as positive or negative with green/red text, which I think is a great idea. So maybe there's a "spin" property within whatever form the endgame condition takes.
- Multiple object states. The need for this is visible in Joe's demo game when you try to elbow-drop the coffee table twice: there's no way for it to say something different the second time, when in the player's mind it's become splinters. My suggestion for this is (I believe) stolen wholesale from ZZT: you allow multiple instances of an item in the script, like so:
  
  ```
  {"Name": "Coffee Table", "ElbowDrop": ["You elbow drop that mess. Wood flies everywhere.", "It's a pile of splinters now, you'll just hurt yourself."]}
  ```

...so if a given item has a verb property that's been assigned a list rather than a flat string, the interpreter knows to increment over to the next member of that list if/when the coffee table gets elbow-dropped. There may be a much more elegant way to do this, but that's one option.

More advanced:
- Counters. Consider a general-purpose way to increment and decrement variables within the story environment. Where we currently have "items" and "room" list properties, let's say we add one called "counters." It looks something like this:

"Counters" : [
{"Name": "applesCollected", "initialValue": 0, "endgame" = true, "endgameValue": 20, "endgameSpin"; "positive"}
]

and then up in "Items":

```
    {
      "Name": "ripening green apple", 
      "Take": "true"
      "takeCounter": ["applesCollected", 1]
    }, 
    {
      "Name": "mouldering emerald apple", 
      "Take": "true"
      "takeCounter": "applesCollected"
    }, 
```

... etc.

This gives you a counting concept that you can use, in this example, to end the game when the player collects 20 apples. There's a counter called applesCollected that the interpreter knows to increment by 1 when an apple is taken. Alternatively you could make a mafia game where the object is to shoot four snitches, and the same principle/affordances of this model would apply.

**Comment by JoeBernardi** [comment]
Completely approve of this stuff. I assume that--at least for a little while--improvements to the JSON spec will take a backseat to bringing the GUI up to speed (working on it!), but victory/defeat states (along with the weirdly complex dynamic of using an item on another item) are a big gap in the current spec. Multiple object states would be a great addition, though, and would be easy to implement. A starting room state would be really easy to add to my current javascript file, too. 

The "counters" idea is great, and pretty similar to the half-baked (and currently pretty much unimplemented) solution I've devised for an item derived from combining two items (e.g., using "sunglasses" on "jackalope" would delete both "sunglasses" and "jackalope" but create "fresh jackalope"). I imagine keeping track of stuff like that in the browser and regularly testing them against constants in the JSON will become de rigueur as the spec gets more and more complex.

**Comment by irondavy** [comment]
Starting room property: of course.

A way to end games: yes, for sure. I don't know where it should live. There's something interesting about attaching it only to arbitrary counters, as you could use counters as AND or OR conditions, effectively.

Also, my point about inflecting an ending with a positive or negative attribute: that was for printing messages in general, not just ending messages.

Multiple object states is useful, and I think the solution you describe is actually really elegant without being complex. It's not going to capture 100% of cases, but I think there will be workarounds for when you need something more complex, like swapping in an object with another, different object instead of changing states.

Basically I'm all for it, the only thing I'm adding here is the idea that endings can only be attached to counters. Any problems with that? I'm looking ahead to the GUI and that's why I'm interested in keep endings in a single place.

**Comment by ardubs** [comment]
Actually, counters would allow us to fake boolean variables (i.e. conditions) by limiting them to a range of 0-1.

**Comment by irondavy** [comment]
Ah, yeah, I had assumed you were implying that from the start.

**Comment by ardubs** [comment]
Actually now that I think about it some more, counters could be a nice way to control item states. Assign a counter by name to an item, and the interpreter will know to use that counter's value as the index of the item state list to access.

This would mean being able to toggle and move between item states non-destructively, i.e. turning a television on and off, or even changing channels on it in sequence and then looping back to the first channel.

**Comment by irondavy** [comment]
How would that be expressed in the JSON? I'm having a hard time picturing it.

**Comment by ardubs** [comment]
I think it'd be something like:

"Name": "The Accomplished Schoolmarm",
"Description": {"roomStates": ["It has waxy branches, and mostly green
fruit.", "It has been entirely stripped of fruit.", "The tree is on
fire.", "The tree is a smoldering ruin."], "indexCounter":
"schoolmarmCounter"}

...and then later you'd define schoolmarmCounter to start at 0. If the
object of the game were to get the Accomplished Schoolmarm to become a
smoldering ruin, you could assign a victory condition to
schoolmarmCounter reaching 3.
pretty inelegant way to set up what is basically an array, but i think
it'd work.

On Tue, Jan 24, 2012 at 11:13 AM, David Cole
reply@reply.github.com
wrote:

**Comment by irondavy** [comment]
How much of that is the way JSON works vs. an original idea for how to do this in Picaro?
